### PR TITLE
Finalizing Binding interfaces for v2.

### PR DESCRIPTION
**Pull request by MikeStall**
This does some major plumbing fixes to ConverterManager and removes the [Obsolete] attributes. 
This has breaking changes that will need to be pushed through other extensions to be on a full BYOB pipeline. 

  https://github.com/Azure/azure-webjobs-sdk/issues/1406
  https://github.com/Azure/azure-webjobs-sdk/issues/686
  https://github.com/Azure/azure-webjobs-sdk/issues/757

1) Adding converters should be done via an extension and standard rules; and not via direct IConverterManager calls.
This lets us shrink public surface area:
- Remove most of BindingFactory and IConverterManager/Extesions.
- some renames to clarify between Exact types and OpenTypes

2) Async converters

3) Stronger invariants:
- Better asserts for misusing open converters
- Have QueueProvider now have an explicit rule for Poco --> QueueMessage (Rather than hackey). Other extensions can use this same Poco definition to get consistent behavior.
- Allow adding multiple rules to an attribute. This was important for factoring and composition.
- fix subtle bugs about the difference between Open Type and Exact Type

4) fixing issues in Converter manager:
- Loose signature using System.Type instead of requiring <T> everywhere. That lets callers avoid needing to MakeGenericType just to call .
- Async everywhere
- Stronger consistency between OpenType and System.Type

### CODE REVIEW

File: src/Microsoft.Azure.WebJobs.Host/Queues/Bindings/QueueBindingProvider.cs [hunk 1]
```diff
@@ -80,19 +81,30 @@ public void Initialize(ExtensionConfigContext context)
 
                 binding.SetPostResolveHook(ToReadWriteParameterDescriptorForCollector)
                         .BindToInput<CloudQueue>(builder);
+            }
+
+            private async Task<IStorageQueueMessage> ConvertPocoToCloudQueueMessage(object arg, Attribute attrResolved, ValueBindingContext context)
+            {
+                var attr = (QueueAttribute)attrResolved;
```
Comment (brettsam) [inline src/Microsoft.Azure.WebJobs.Host/Queues/Bindings/QueueBindingProvider.cs]: Since this hangs off of a binding with known TAttribute, can we make it pass TAttribute to the func? Then you wouldn't have to cast.

File: src/Microsoft.Azure.WebJobs.Host/Bindings/BindingFactory.cs [hunk 2]
```diff
@@ -33,15 +33,15 @@ public BindingFactory(INameResolver nameResolver, IConverterManager converterMan
         /// <summary>
         /// Get the name resolver for resolving %% tokens. 
         /// </summary>
-        public INameResolver NameResolver
+        internal INameResolver NameResolver
```
Comment (brettsam) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/BindingFactory.cs]: Why are these moved to internal and not just removed?

File: src/Microsoft.Azure.WebJobs.Host/Bindings/PatternMatcher.cs [hunk 3]
```diff
@@ -268,6 +268,7 @@ private static bool CheckArg(Type openType, Type specificType, Dictionary<string
         // Helper for getting a converter func that invokes the converter interface on an object.
         private abstract class InvokerBase
         {
+            // This should get changed to a FuncAsyncConverter
```
Comment (brettsam) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/PatternMatcher.cs]: Can you make that change here?

File: src/Microsoft.Azure.WebJobs.Host/Config/FluentConverterRules.cs [hunk 4]
```diff
@@ -16,7 +17,7 @@ namespace Microsoft.Azure.WebJobs.Host.Config
     public abstract class FluentConverterRules<TAttribute, TThis> where TAttribute : Attribute
     {
         // Access the converter manager that we're adding rules to. 
-        internal abstract IConverterManager Converters { get; }
+        internal abstract ConverterManager Converters { get; }
```
Comment (brettsam) [inline src/Microsoft.Azure.WebJobs.Host/Config/FluentConverterRules.cs]: Why not use the interface?

File: src/Microsoft.Azure.WebJobs.Host/Bindings/OpenType.cs [hunk 5]
```diff
@@ -6,28 +6,328 @@
 
 namespace Microsoft.Azure.WebJobs.Host.Bindings
 {
+    // Ensures that open Types match consistently 
+    public class OpenTypeMatchContext
```
Comment (brettsam) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/OpenType.cs]: Should this be public?

File: src/Microsoft.Azure.WebJobs.Host/Bindings/BindingFactory.cs (hunk 2 above)
Comment (MikeStall) [reply src/Microsoft.Azure.WebJobs.Host/Bindings/BindingFactory.cs]: They're still used by the Fluent rules (from ExtensionConfigContext).  Those rules are the public surface area that people should be using. 

---
In reply to: [148657335](https://github.com/Azure/azure-webjobs-sdk/pull/1413#discussion_r148657335) [](ancestors = 148657335)

File: src/Microsoft.Azure.WebJobs.Host/Config/FluentConverterRules.cs (hunk 4 above)
Comment (MikeStall) [reply src/Microsoft.Azure.WebJobs.Host/Config/FluentConverterRules.cs]: The public interface is very constrained (just a Getter). Whereas the class is private, and it has additional setters. I'm trying to shirnk the public surface area as much as possible to reduce liability. 

---
In reply to: [148663519](https://github.com/Azure/azure-webjobs-sdk/pull/1413#discussion_r148663519) [](ancestors = 148663519)

File: src/Microsoft.Azure.WebJobs.Host/Bindings/OpenType.cs (hunk 5 above)
Comment (MikeStall) [reply src/Microsoft.Azure.WebJobs.Host/Bindings/OpenType.cs]: Unfortunately, it needs to be since public abstract OpenType.IsMatch(Type type, OpenTypeMatchContext context); needs to be public. 
That method needs to be public since derived classes may implement it. But the only thing a derived class would do with this context is pass it into a IsMatch call on one of our builtin opentypes. 

---
In reply to: [148666948](https://github.com/Azure/azure-webjobs-sdk/pull/1413#discussion_r148666948) [](ancestors = 148666948)

File: src/Microsoft.Azure.WebJobs.Host/Queues/Bindings/QueueBindingProvider.cs (hunk 1 above)
Comment (MikeStall) [reply src/Microsoft.Azure.WebJobs.Host/Queues/Bindings/QueueBindingProvider.cs]: I'll try - the problem I'm running into is "delegate overload hell".  We have too many permutations, and it stresses C#'s overload rules. 

---
In reply to: [148656706](https://github.com/Azure/azure-webjobs-sdk/pull/1413#discussion_r148656706) [](ancestors = 148656706)

**Comment by MikeStall** [comment]
This has been superseded by: https://github.com/Azure/azure-webjobs-sdk/pull/1422

**Comment by MikeStall** [comment]
This has been superseded by: https://github.com/Azure/azure-webjobs-sdk/pull/1422
